### PR TITLE
Fixes and Improvements for multi-node test runs

### PR DESCRIPTION
**Pull request by devreal**
I tried to run our test suite on multiple nodes on our cluster and came across some bugs that don't seem to manifest as errors if running on one node only. In particular, this PR contains fixes to the handling of `MPI_Request` objects and an issue in our shared memory optimization. 

Additionally, this PR adds `dash::generate_with_index` to enable parallel generation of container values in cases where the generated value depends on the index of the element. This function is applied to two tests in which the serial generation was really slow on 8 nodes.

### CODE REVIEW

File: dart-impl/mpi/src/dart_communication.c [hunk 1]
```diff
@@ -1276,40 +1278,33 @@ dart_ret_t dart_waitall_local(
...
-            if (MPI_Request_free(&mpi_req[r_n]) != MPI_SUCCESS) {
-              DART_LOG_TRACE("dart_waitall_local ! MPI_Request_free failed");
-              ret = DART_ERR_INVAL;
-              break;
-            }
-          } else {
-            DART_LOG_TRACE("dart_waitall_local: cannot free request %zu "
-                           "mpi_sta[%zu] = %d (%s)",
-                           r_n,
+        if (handle[i]->request) {
```
Comment (fmoessbauer) [inline dart-impl/mpi/src/dart_communication.c:1285]: I am surprised that this even worked in the past...

File: dart-impl/mpi/src/dart_globmem.c [hunk 2]
```diff
@@ -283,7 +283,7 @@ dart_team_memalloc_aligned(
   MPI_Comm_rank(sharedmem_comm, &sharedmem_unitid);
   // re-use previously allocated memory
   if (segment->baseptr == NULL) {
-    segment->baseptr = malloc(sizeof(char *) * team_data->sharedmem_nodesize);
+    segment->baseptr = calloc(team_data->sharedmem_nodesize, sizeof(char *));
```
Comment (fmoessbauer) [inline dart-impl/mpi/src/dart_globmem.c:286]: `calloc` should not be necessary here, but who cares ;)

File: dash/include/dash/algorithm/Generate.h [hunk 3]
```diff
@@ -48,6 +48,55 @@ void generate (
...
+ * \complexity  O(d) + O(nl), with \c d dimensions in the global iterators'
+ *              pattern and \c nl local elements within the global range
+ *
+ * \ingroup     DashAlgorithms
+ */
+template <
+    typename ElementType,
+    class    PatternType,
+    class    UnaryFunction >
+void generate_with_index(
```
Comment (fmoessbauer) [inline dash/include/dash/algorithm/Generate.h:73]: Cool, I missed this function quite often.

File: dash/include/dash/algorithm/Generate.h [hunk 4]
```diff
@@ -48,6 +48,55 @@ void generate (
...
+  /// Iterator to the initial position in the sequence
+  GlobIter<ElementType, PatternType> first,
+  /// Iterator to the final position in the sequence
+  GlobIter<ElementType, PatternType> last,
+  /// Generator function
+  UnaryFunction                      gen) {
+  /// Global iterators to local index range:
+  auto index_range  = dash::local_index_range(first, last);
```
Comment (fmoessbauer) [inline dash/include/dash/algorithm/Generate.h:81]: Better use the new range based views, as this does not work with multiple local ranges AFAIK. Maybe @fuchsto can assist.

File: dash/include/dash/algorithm/Generate.h [hunk 5]
```diff
@@ -48,6 +48,55 @@ void generate (
...
+  auto lbegin_index = index_range.begin;
+  auto lend_index   = index_range.end;
+
+  if (lbegin_index != lend_index) {
+    // Pattern from global begin iterator:
+    auto & pattern    = first.pattern();
+    auto first_offset = first.pos();
+    // Iterate local index range:
+    for (auto lindex = lbegin_index;
```
Comment (fmoessbauer) [inline dash/include/dash/algorithm/Generate.h:90]: Same problem here.

**Review by fmoessbauer — COMMENTED** [review]
Thanks for the fixes. See my comments inline.

File: dash/include/dash/algorithm/Generate.h (hunk 4 above)
Comment (devreal) [reply dash/include/dash/algorithm/Generate.h:81]: Do you have any pointer on how to use them? I'm not sure range based views and I have crossed paths so far...

Comment (fuchsto) [reply dash/include/dash/algorithm/Generate.h:81]: Hey, I even introduced you personally last time in Munich!
Anyways: yes, I will add documentation.
For usage you can refer to quite an abundance of tests in dash/test/view.

Comment (devreal) [reply dash/include/dash/algorithm/Generate.h:81]: To be honest, I don't even know where to start looking. I guess I could crawl through a host of undocumented code, still not knowing where to start or how to put the pieces together...

**Review by fuchsto — APPROVED** [review]

**Pull request merged**